spec.bs

```diff
@@ -324,13 +325,22 @@ Moreover, each {{SharedStorageWorklet}}'s [=global scopes|list of global scopes=
   ## Run Operation Methods on {{SharedStorageWorklet}} ## {#run-op-shared-storage-worklet}
 
   <div algorithm>
-    To <dfn>get the select-url result index</dfn>, given {{SharedStorageWorklet}} |worklet|, {{DOMString}} |operationName|, [=/list=] of {{SharedStorageUrlWithMetadata}}s |urlList|, an [=/origin=] |workletDataOrigin|, {{SharedStorageRunOperationMethodOptions}} |options|, a [=pre-specified report parameters=] or null |preSpecifiedParams| and an [=aggregation coordinator=] or null |aggregationCoordinator|:
+    To <dfn>get the select-url result index</dfn>, given {{SharedStorageWorklet}} |worklet|, {{DOMString}} |operationName|, [=/list=] of {{SharedStorageUrlWithMetadata}}s |urlList|, an [=/origin=] |workletDataOrigin|, a [=/navigable=] |navigable|, {{SharedStorageRunOperationMethodOptions}} |options|, a [=pre-specified report parameters=] or null |preSpecifiedParams| and an [=aggregation coordinator=] or null |aggregationCoordinator|:
 
     1. Let |promise| be a new [=promise=].
     1. Let |window| be |worklet|'s [=relevant settings object=].
     1. [=Assert=]: |window| is a {{Window}}.
     1. If |window|'s [=Window/browsing context=] is null, then return a [=promise rejected=] with a {{TypeError}}.
     1. If |window|'s [=associated document=] is not [=fully active=], return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |savedQueryName| be |options|["`savedQuery`"].
+    1. If |savedQueryName| is nonempty, then:
+        1. Let |site| be the result of running [=obtain a site=] on |workletDataOrigin|.
+        1. Let |savedIndex| be the result of running [=get the index for a saved query=] on |navigable|, |site|, and |savedQueryName|.
```

Comment:
It appears that you could use the same name for two completely different queries and this would return the value of the first query.  Is this the desired behavior?

For example, maybe we should return an error if the parameters passed to selectUrl in the second call differ from the first call?  Or should the second call override the parameters from the first selectUrl by creating a new saved query?

This is kind of stemming from the awkardness of adding a parameter to an existing function call.  It seems like what you might want is something more like how SQL has separate "prepare statement" and "execute statement" calls.  You could consider a `prepareSelectUrl()` that takes all params and assigns a name.  Then a separate `selectUrl(name)` call that operates only on the name without any other params.

If we want to keep a single function call that is overloaded I think we should just be intentional if the site calls it twice with different arguments, but the same name.

Reply:
> It appears that you could use the same name for two completely different queries and this would return the value of the first query. Is this the desired behavior?

The short answer is: probably yes.  

The long answer is that I did forget to take into account the operation name, and I may not want to ignore that. I am thinking about that some more offline.  

But we deliberately want the array of URLs to be able to be different, and note that we do handle the error that happens if we end up with an out of range index because of this.

We would ignore any passed `data` for `options` in the case that we are able to retrieve a saved query result. The other `options` parameters are still handled.

So there shouldn't be a need for a separate `prepareSelectUrl()` statement.

Reply:
In that case I recommend returning an error if the operation name doesn't match the previously used operation name for the same saved query.  Its probably a developer mistake.  Allowing the same saved query name for different operations sounds like a footgun.

Reply:
After further contemplation and discussion with the team, we've decided two things.

1. Saved queries will be stored on a per-origin (instead of per-site) basis. I have updated this spec PR accordingly.
2. We do indeed want names of saved queries to be independent of operation names, and for the storing and lookup of saved queries to ignore the operation names. So two calls to `selectURL()` that have different operation names could still access the same saved query result if they use the same `savedQuery` parameter. While this could lead to some developers making mistakes, it also allows for greater flexibility and simplicity. We would need to be sure to document this properly, of course, including in the current explainer PR (where the effect of differing operation names is currently not mentioned).

Reply:
I don't have particularly strong feelings here. It seems unlikely to be a footgun to me, but at the same time I can't think of a good reason to use the same saved-query on two different operations. So, let's say that we go ahead with your proposal here as it seems reasonable.

So now the proposal is that a saved-query key is `<origin, operation_name, saved_query_name>`. 

For similar reasons, what would you think about also including the worklet script url as part of the key?

e.g., `<origin, operation_name, worklet_script_url, saved_query_name>`

Like operations, presumably different scripts could have different behavior, even if the operation name is the same. This gets a little funky in that urls aren't actually const resources, but still.

Reply:
OK, I have updated the spec so that queries are keyed by `<origin, operation_name, saved_query_name>`.

I think we should just not reuse the query if there is different operation name, rather than throwing, to simplify things.

Reply:
Just to be clear, keying on `<origin, operation_name, saved_query_name>` does your proposal above, right?  It ignores mismatched operations because you end up with different slots in your map.  To do what I was proposing I think the key needs to be `<origin, saved_query_name>` and then in the map value the `operation_name` is saved so you can compare it on lookup.

I'm not sure I understand enough about how different worklets can mix on a page to have an opinion on how to treat conflicting scripts URLs.  If we assume folks are going to hash URLs for different versions of the same worklet, though, it does seem like a flow we should be concerned with.

Reply:
> Just to be clear, keying on <origin, operation_name, saved_query_name> does your proposal above, right? It ignores mismatched operations because you end up with different slots in your map. To do what I was proposing I think the key needs to be <origin, saved_query_name> and then in the map value the operation_name is saved so you can compare it on lookup.

Yes. The key would need to be `<origin, saved_query_name` to do what you propose. But I think it would be best to key on `<origin, operation_name, saved_query_name>` instead because it seems to make more sense for us to simply narrow the scope for the saved queries further from the beginning, then to add additional errors. We wouldn't throw if there is a mismatch, but we wouldn't re-use a query in the case of a mismatch either.